external-crates/move/crates/move-package-alt/src/graph/package_info.rs

```diff
@@ -182,43 +213,105 @@ impl<'graph, F: MoveFlavor> PackageInfo<'graph, F> {
 
     /// For legacy packages, our named addresses need to include all transitive deps too.
     /// An example of that is depending on "sui", but also keeping it possible to use "std".
-    fn legacy_named_addresses(&self) -> PackageResult<BTreeMap<PackageName, NamedAddress>> {
+    fn legacy_named_addresses(
+        &self,
+        root_assignments: &BTreeMap<PackageName, NamedAddress>,
+    ) -> PackageResult<BTreeMap<PackageName, NamedAddress>> {
         let mut result: BTreeMap<PackageName, NamedAddress> = BTreeMap::new();
 
-        // We only add the package name if it is not the special "unnamed" package name
+        // Step 1: Handle the "name" of the package.
+        // For legacy cases that the name was `_`, there are these options:
+        // 1. The package has an original id defined for the env, in which case thats the correct address
+        // 2. The parent has defined an ID for the address, so we can use that
+        // 3. If none of the above, we fail.
         if self.package().name().as_str() != NO_NAME_LEGACY_PACKAGE_NAME {
-            result.insert(self.package().name().clone(), self.node_to_addr(self.node));
-        }
-
-        for (_, dep) in self.direct_deps() {
-            let transitive_result = dep.legacy_named_addresses()?;
-
-            for (name, addr) in transitive_result {
-                let existing = result.insert(name.clone(), addr.clone());
-
-                if existing.is_some_and(|existing| existing != addr) {
-                    return Err(PackageError::DuplicateNamedAddress {
-                        address: name,
+            // If the package has an `_` as its address, we need special handling
+            if let Some(legacy_data) = &self.package().legacy_data
+                && legacy_data.modern_name_address_is_underscore
+            {
+                // if we have an original id, we do the typical insertion
+                if self.package().original_id().is_some() {
+                    result.insert(self.package().name().clone(), self.node_to_addr(self.node));
```

Comment:
need to handle case where address is defined in the publish file and in the parent - needs to be consistent